compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1717,58 +1717,72 @@ class Typer extends Namer
+            case _ =>
+              val localContext = ctx.localContext(tree, sym)
+              tree match {
+                case tree: untpd.ValDef =>
+                  if (tree.isEmpty) tpd.EmptyValDef
+                  else typedValDef(tree, sym)(localContext.setNewScope)
+                case tree: untpd.DefDef =>
+                  val typer1 = localTyper(sym)
+                  typer1.typedDefDef(tree, sym)(localContext.setTyper(typer1))
+                case tree: untpd.TypeDef =>
+                  if (tree.isClassDef)
+                    typedClassDef(tree, sym.asClass)(localContext.setMode(ctx.mode &~ Mode.InSuperCall))
+                  else
+                    typedTypeDef(tree, sym)(localContext.setNewScope)
+              }
           }
         }
 
         def typedUnnamed(tree: untpd.Tree): Tree = tree match {
-          case tree: untpd.Apply =>
-            if (ctx.mode is Mode.Pattern) typedUnApply(tree, pt) else typedApply(tree, pt)
-          case tree: untpd.This => typedThis(tree)
-          case tree: untpd.Literal => typedLiteral(tree)
-          case tree: untpd.New => typedNew(tree, pt)
-          case tree: untpd.Typed => typedTyped(tree, pt)
-          case tree: untpd.NamedArg => typedNamedArg(tree, pt)
-          case tree: untpd.Assign => typedAssign(tree, pt)
-          case tree: untpd.Block => typedBlock(desugar.block(tree), pt)(ctx.fresh.setNewScope)
-          case tree: untpd.If => typedIf(tree, pt)
-          case tree: untpd.Function => typedFunction(tree, pt)
-          case tree: untpd.Closure => typedClosure(tree, pt)
-          case tree: untpd.Import => typedImport(tree, retrieveSym(tree))
-          case tree: untpd.Match => typedMatch(tree, pt)
-          case tree: untpd.Return => typedReturn(tree)
-          case tree: untpd.Try => typedTry(tree, pt)
-          case tree: untpd.Throw => typedThrow(tree)
-          case tree: untpd.TypeApply => typedTypeApply(tree, pt)
-          case tree: untpd.Super => typedSuper(tree, pt)
-          case tree: untpd.SeqLiteral => typedSeqLiteral(tree, pt)
-          case tree: untpd.Inlined => typedInlined(tree, pt)
-          case tree: untpd.Quote => typedQuote(tree, pt)
-          case tree: untpd.TypeTree => typedTypeTree(tree, pt)
+          case tree: untpd.Apply             => if (ctx.mode is Mode.Pattern) typedUnApply(tree, pt) else typedApply(tree, pt)
+          case tree: untpd.This              => typedThis(tree)
+          case tree: untpd.Literal           => typedLiteral(tree)
+          case tree: untpd.New               => typedNew(tree, pt)
+          case tree: untpd.Typed             => typedTyped(tree, pt)
+          case tree: untpd.NamedArg          => typedNamedArg(tree, pt)
+          case tree: untpd.Assign            => typedAssign(tree, pt)
+          case tree: untpd.Block             => typedBlock(desugar.block(tree), pt)(ctx.fresh.setNewScope)
+          case _ => typedUnnamed1(tree)
```

Comment:
I doubt this kind of "outlining" is useful since I expect that the JIT will just inline it right back.

Reply:
I pushed the last commit (doing the outlining stuff) by mistake, I'm started some benchmarks locally to see how this helps...